pdfminer/converter.py

```diff
@@ -76,49 +76,46 @@ def paint_path(self, gstate, stroke, fill, evenodd, path):
-                          fill, evenodd, gstate.scolor, gstate.ncolor)
-            self.cur_item.add(rect)
-
-        # if path contains multiple subpaths, split them up and reprocess
-        elif shape.count('m') > 1:
+        if shape.count('m') > 1:
+            # recurse if there are multiple m's in this shape
             for m in re.finditer(r'm[^m]+', shape):
                 subpath = path[m.start(0):m.end(0)]
                 self.paint_path(gstate, stroke, fill, evenodd, subpath)
 
-        elif shape == 'ml':
-            # single line segment
-            (x0, y0) = apply_matrix_pt(self.ctm, path[0][1:])
-            (x1, y1) = apply_matrix_pt(self.ctm, path[1][1:])
-            if x0 == x1 or y0 == y1:
-                line = LTLine(gstate.linewidth, (x0, y0), (x1, y1), stroke,
-                              fill, evenodd, gstate.scolor, gstate.ncolor)
-                self.cur_item.add(line)
-
         else:
-            # if not a rectangle or a line, treat as a curve
-            pts = []
-            for p in path:
-                for i in range(1, len(p), 2):
-                    pts.append(apply_matrix_pt(self.ctm, (p[i], p[i+1])))
-            curve = LTCurve(gstate.linewidth, pts, stroke, fill, evenodd,
-                            gstate.scolor, gstate.ncolor)
-            self.cur_item.add(curve)
+            if shape == 'ml':
+                # single line segment
+                (x0, y0) = apply_matrix_pt(self.ctm, path[0][1:])
+                (x1, y1) = apply_matrix_pt(self.ctm, path[1][1:])
+                if x0 == x1 or y0 == y1:
+                    line = LTLine(gstate.linewidth, (x0, y0), (x1, y1), stroke,
+                                  fill, evenodd, gstate.scolor, gstate.ncolor)
+                    self.cur_item.add(line)
+
+            elif shape == 'mlllh':
+                (x0, y0) = apply_matrix_pt(self.ctm, path[0][1:])
+                (x1, y1) = apply_matrix_pt(self.ctm, path[1][1:])
+                (x2, y2) = apply_matrix_pt(self.ctm, path[2][1:])
+                (x3, y3) = apply_matrix_pt(self.ctm, path[3][1:])
+
+                if (x0 == x1 and y1 == y2 and x2 == x3 and y3 == y0) or \
+                        (y0 == y1 and x1 == x2 and y2 == y3 and x3 == x0):
+                    rect = LTRect(gstate.linewidth, (x0, y0, x2, y2), stroke,
+                                  fill, evenodd, gstate.scolor, gstate.ncolor)
+                    self.cur_item.add(rect)
+                else:
+                    curve = self._create_curve(gstate, stroke, fill, evenodd, path)
+                    self.cur_item.add(curve)
+
+            else:
+                curve = self._create_curve(gstate, stroke, fill, evenodd, path)
+                self.cur_item.add(curve)
+
+    def _create_curve(self, gstate, stroke, fill, evenodd, path):
+        """Create a `LTCurve` object for the paint path operator"""
+        pts = [apply_matrix_pt(self.ctm, point) for p in path for point in zip(p[::2], p[1::2])]
```

Comment:
`tests/test_converter.py:test_paint_path_quadrilaterals` appears to be failing on this line.

Reply:
I'll look into it

Reply:
I made an error when restructuring the for loops into a list comprehension. It should be `zip(p[1::2], p[2::2])`.

Reply:
The new tests are already proving their value :D